commons/commons-serialization/src/main/kotlin/ac/uk/ebi/biostd/tsv/deserialization/common/Functions.kt

```diff
@@ -18,56 +17,51 @@ internal inline fun validate(value: Boolean, lazyMessage: () -> String) {
-    val attributes: MutableList<Attribute> = mutableListOf()
-    chunkLines.forEach { line ->
-        when {
-            line.isNameDetail() -> addNameAttributeDetail(line.name(), line.value!!, attributes)
-            line.isValueDetail() -> addValueAttributeDetail(line.name(), line.value!!, attributes)
-            else -> attributes.add(Attribute(line.name(), line.value.nullIfBlank(), line.isReference()))
-        }
-    }
-    return attributes
-}
+internal fun toAttributes(chunkLines: List<TsvChunkLine>): List<Attribute> =
+    getAttributes(chunkLines.map { it.name to it.value.nullIfBlank() })
 
-internal fun <T> asTable(chunk: TsvChunk, initializer: (String, MutableList<Attribute>) -> T): List<T> {
-    val rows: MutableList<T> = mutableListOf()
-
-    chunk.lines.ifEmpty { throw InvalidElementException(REQUIRED_TABLE_ROWS) }
-    chunk.lines.forEach {
-        val attrs: MutableList<Attribute> = mutableListOf()
-        val rowAttrsSize = it.rawValues.size
-        val headerAttrsSize = chunk.header.size - 1
-
-        validate(rowAttrsSize <= headerAttrsSize) { throw InvalidElementException(INVALID_TABLE_ROW) }
-
-        it.rawValues.forEachIndexed { index, attr ->
-            attr.applyIfNotBlank { parseTableAttribute(chunk.header[index + 1], attr, attrs) }
+internal fun <T> asTable(chunk: TsvChunk, initializer: (String, List<Attribute>) -> T): List<T> {
+    val rows = buildList {
+        chunk.lines.ifEmpty { throw InvalidElementException(REQUIRED_TABLE_ROWS) }
+        chunk.lines.forEach { line ->
+            val attrs = getAttributes(line, chunk)
+            add(initializer(line.name(), attrs))
         }
-
-        rows.add(initializer(it.name(), attrs))
     }
 
-    return rows.toList()
+    return rows
 }
 
-private fun parseTableAttribute(name: String, value: String, attributes: MutableList<Attribute>) {
-    when {
-        isNameDetail(name) -> addNameAttributeDetail(getDetailName(name), value, attributes)
-        isValueDetail(name) -> addValueAttributeDetail(getDetailName(name), value, attributes)
-        else -> attributes.add(Attribute(name, value))
-    }
+private fun getAttributes(line: TsvChunkLine, chunk: TsvChunk): List<Attribute> {
+    validate(line.size <= chunk.header.size) { throw InvalidElementException(INVALID_TABLE_ROW) }
+    val values = chunk.header.rawValues.mapIndexed { i, value -> value to line.rawValues.getOrNull(i).nullIfBlank() }
+    return getAttributes(values)
 }
 
-private fun getDetailName(attrName: String) = attrName.substring(1, attrName.lastIndex)
-
-private fun addNameAttributeDetail(name: String, value: String, attributes: MutableList<Attribute>) {
-    attributes.ifEmpty { throw InvalidElementException(MISPLACED_ATTR_NAME) }
-    attributes.last().nameAttrs.add(AttributeDetail(name, value))
+private fun getAttributes(values: List<Pair<String, String?>>): List<Attribute> = buildList {
+    var previous: Attribute? = null
+    for ((header, value) in values) {
```

Comment:
Just to clarify: we are only accepting empty values for general attributes, right? Neither namequal nor valqual

Reply:
good question let me ask question to Ugis as I do not know.